### DISCUSSION TITLE
Cant specify who can force push via github api or gh !!

### DISCUSSION THREAD

**Discussion by sayyedmooaz** [community/community#148166] [Apps, API and Webhooks]
### Select Topic Area

Show & Tell

### Body

i Want to setup some branch protection rules and want to specify who can force push but github api only accepts boolean values.
What I am trying to do 
<img width="710" alt="Screenshot 2024-12-30 at 4 28 07 PM" src="https://github.com/user-attachments/assets/027e8611-03a6-479d-a704-f5ee92880aa2" />
This is very important in my workflow , either i need an api or cmd that can do this , can anyone help i want to integrate this step for a pipeline.

**Comment by rmandvikar** [comment]
same issue: https://github.com/orgs/community/discussions/128770

**Comment by davex-ai** [comment]
To specify which users or teams can force-push via the GitHub API, you must use the bypass_force_push_allowances parameter within the "Update branch protection" endpoint.
While the standard allow_force_pushes parameter is a simple boolean, [GitHub](https://github.com/marketplace/actions/push-to-status-check-protected-branches) introduced granular force-push controls that require a different object structure. 

API Implementation
To integrate this into your pipeline, use the following REST API structure:
Endpoint: PUT /repos/{owner}/{repo}/branches/{branch}/protection
Request Body (Partial):
json
{
  "allow_force_pushes": true,
  "bypass_force_push_allowances": {
    "users": ["octocat"],
    "teams": ["my-team"],
    "apps": []
  }
}


allow_force_pushes: Must be set to true to enable the feature.
bypass_force_push_allowances: This object specifies the actors (users, teams, or apps) that are allowed to force-push even when protection is active. 

Key Requirements
Permissions: You must have admin permissions on the repository or a custom role with "edit repository rules".
Organization Plans: Specifying teams or users for force-push is typically available for organizations using GitHub Team or GitHub Enterprise Cloud, and for public repositories.
Admin Enforcement: By default, branch protection rules do not apply to administrators unless you also enable the enforce_admins setting in your API call.
Upvote if this helps

**Comment by rmandvikar** [comment]
`bypass_pull_request_allowances` is NOT what affects the "Specify who can force push" setting from GH UI. `allow_force_pushes` as `true` will default to "Everyone".
